htlcswitch/switch.go

```diff
@@ -547,35 +541,23 @@ func (s *Switch) htlcForwarder() {
 
 	for {
 		select {
-		case req := <-s.chanCloseRequests:
-			s.handleChanelClose(req)
-
-		case cmd := <-s.htlcPlex:
-			var (
-				paymentHash lnwallet.PaymentHash
-				amount      btcutil.Amount
-			)
-
-			switch m := cmd.pkt.htlc.(type) {
-			case *lnwire.UpdateAddHTLC:
-				paymentHash = m.PaymentHash
-				amount = m.Amount
-			case *lnwire.UpdateFufillHTLC, *lnwire.UpdateFailHTLC:
-				paymentHash = cmd.pkt.payHash
-				amount = cmd.pkt.amount
-			default:
-				cmd.err <- errors.New("wrong type of update")
-				return
-			}
-
-			payment, err := s.findPayment(amount, paymentHash)
-			if err != nil {
-				cmd.err <- s.handlePacketForward(cmd.pkt)
-			} else {
-				cmd.err <- s.handleLocalDispatch(payment, cmd.pkt)
+		case req := <-s.linkControl:
+			switch cmd := req.(type) {
+			case *addLinkCmd:
+				cmd.err <- s.addLink(cmd.link)
```

Comment:
It seems that we'll need to introduce mutexes internally for this new set of changes to be concurrent safe.